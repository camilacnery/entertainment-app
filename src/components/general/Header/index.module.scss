@use '@/styles/constants' as *;

.header {
  display: flex;
  padding: 3rem 2rem;
  align-items: center;
  justify-content: space-between;
  background-color: $background-box-color;
  box-shadow: $shadow;
  height: 8vh;
  z-index: 3;
  font-weight: bold;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.leftContent {
  display: flex;
  align-items: center;

  h1 {
    color: $primary-color;
    font-family: "Bebas Neue", serif;
    font-size: 3rem;
  }

  @media screen and (max-width: 767px) {
    h1 {
      font-size: 2.5rem;
      margin-top: 2rem;
    }

    nav {
      display: none;
    }
  }
}

.user {
  cursor: pointer;
  display: flex;
  justify-content: center;
  align-items: center;

  @media screen and (max-width: 767px) {
    span {
      display: none;
    }
  }

  @media screen and (min-width: 768px) {
    span {
      margin-right: 1rem;
    }
  }
}