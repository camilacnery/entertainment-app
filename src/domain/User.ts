export interface IUserProfile {
  imageUrl: string;
  fullName: string;
  firstName: string;
  email: string;
}
