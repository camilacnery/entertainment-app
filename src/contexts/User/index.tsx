import { ReactNode, FC, createContext, useState, useContext, useEffect } from "react";
import { IUserProfile } from "@/domain/User";
import { decodeJwt } from "@/utils/jwt";

interface IDecodedUser {
  aud: string;
  azp: string;
  email: string;
  email_verified: boolean;
  exp: number;
  family_name: string;
  given_name: string;
  iat: number;
  iss: string;
  jti: string;
  name: string;
  nbf: number;
  picture: string;
  sub: string;
}

export type TUserContext = {
  profile?: IUserProfile;
  onSuccessLogin: (userJwt: string) => void;
  onLogout: () => void;
};

const UserContext = createContext<TUserContext>({} as TUserContext);

type Props = {
  children: ReactNode;
};

const STORAGE_ID = "app/user-token";

const UserContextProvider: FC<Props> = ({ children }) => {
  const [profile, setProfile] = useState<IUserProfile>();

  const setProfileFromToken = (userJwt: string) => {
    const profileData = decodeJwt<IDecodedUser>(userJwt);

    setProfile({
      fullName: profileData.name,
      firstName: profileData.given_name,
      email: profileData.email,
      imageUrl: profileData.picture,
    });
  };

  const onSuccessLogin = (userJwt: string) => {
    localStorage.setItem(STORAGE_ID, userJwt);
    setProfileFromToken(userJwt);
  };

  const onLogout = () => {
    localStorage.removeItem(STORAGE_ID);
    setProfile(undefined);
  };

  useEffect(() => {
    const userJwt = localStorage.getItem(STORAGE_ID);
    if (!userJwt) return;

    setProfileFromToken(userJwt);
  }, []);

  return (
    <UserContext.Provider
      value={{
        profile,
        onSuccessLogin,
        onLogout,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

const useUserContext = () => useContext(UserContext);

export { useUserContext, UserContext, UserContextProvider };
