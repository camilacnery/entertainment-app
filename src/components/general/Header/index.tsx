import { FC } from "react";
import Link from "next/link";
import { useUserContext } from "@/contexts/User";
import Avatar from "../Avatar";
import styles from "./index.module.scss";
import Button from "../Button";

const Header: FC = () => {
  const { profile } = useUserContext();

  return (
    <header className={styles.header}>
      <div className={styles.leftContent}>
        <Link href="/">
          <h1>The Catalog</h1>
        </Link>
        <nav>
          <ul>
            <li>
              <Link href="/">Home</Link>
            </li>
          </ul>
        </nav>
      </div>

      <div>
        {profile ? (
          <Link href="/profile">
            <div className={styles.user}>
              <span>{profile.firstName}</span>
              <Avatar size={40} imageUrl={profile.imageUrl} />
            </div>
          </Link>
        ) : (
          <Link href="/login">
            <Button>Login</Button>
          </Link>
        )}
      </div>
    </header>
  );
};

export default Header;
