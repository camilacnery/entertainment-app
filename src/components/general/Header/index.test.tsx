import { render, screen } from "@testing-library/react";
import { TUserContext, UserContext } from "@/contexts/User";

import Header from ".";
import userEvent from "@testing-library/user-event";

const mockedRouterPush = jest.fn();
jest.mock("next/router", () => ({
  useRouter: () => ({
    push: mockedRouterPush,
  }),
}));

const componentElements = {
  loginButton: () => screen.getByRole("button", { name: "Login" }),
};

const renderHeader = (props: Partial<TUserContext> = {}) => {
  return render(
    <UserContext.Provider value={{ onSuccessLogin: jest.fn(), onLogout: jest.fn(), ...props }}>
      <Header />
    </UserContext.Provider>
  );
};

describe("Header", () => {
  it("renders component when user is logged out", () => {
    const { baseElement } = renderHeader();

    expect(baseElement).toMatchSnapshot();
  });

  it("renders component when user is logged in", () => {
    const { baseElement } = renderHeader({
      profile: {
        firstName: "Test",
        fullName: "Test User",
        email: "test@test.com",
        imageUrl: "https://image.url",
      },
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("redirects to login page when login button is clicked", async () => {
    renderHeader();

    await userEvent.click(componentElements.loginButton());

    expect(mockedRouterPush).toHaveBeenCalledWith("/login");
  });
});
