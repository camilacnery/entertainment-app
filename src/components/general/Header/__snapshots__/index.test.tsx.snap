// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders component when user is logged in 1`] = `
<body>
  <div>
    <header
      class="header"
    >
      <div
        class="leftContent"
      >
        <h1>
          The Catalog
        </h1>
        <nav>
          <ul>
            <li>
              <a
                href="/"
              >
                Home
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <div>
        <div
          class="user"
        >
          <span>
            Test
          </span>
          <img
            alt=""
            class="avatar"
            data-nimg="future"
            decoding="async"
            height="40"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fimage.url&w=96&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fimage.url&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.url&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </div>
      </div>
    </header>
  </div>
</body>
`;

exports[`Header renders component when user is logged out 1`] = `
<body>
  <div>
    <header
      class="header"
    >
      <div
        class="leftContent"
      >
        <h1>
          The Catalog
        </h1>
        <nav>
          <ul>
            <li>
              <a
                href="/"
              >
                Home
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <div>
        <button
          class="button"
        >
          Login
        </button>
      </div>
    </header>
  </div>
</body>
`;
